src/Stratis.Bitcoin.Features.PoA/BasePoAFeatureConsensusRules/PoAHeaderSignatureRule.cs

```diff
@@ -29,10 +41,28 @@ public override void Initialize()
             this.slotsManager = engine.SlotsManager;
             this.federationHistory = engine.FederationHistory;
             this.validator = engine.PoaHeaderValidator;
+
+            var lastCheckPoint = engine.Network.Checkpoints.LastOrDefault();
+            this.lastCheckPoint = (lastCheckPoint.Value != null) ? new HashHeightPair(lastCheckPoint.Value.Hash, lastCheckPoint.Key) : null;
         }
 
         public override async Task RunAsync(RuleContext context)
         {
+            // Only start validating at the last checkpoint block.
+            if (this.initialBlockDownloadState.IsInitialBlockDownload() && context.ValidationContext.ChainedHeaderToValidate.Height <= (this.lastCheckPoint?.Height ?? 0))
+            {
+                if (context.ValidationContext.ChainedHeaderToValidate.Height < this.lastCheckPoint.Height)
+                    return;
+
+                // Ensure we're getting the right block at the last checkpoint height.
+                if (context.ValidationContext.ChainedHeaderToValidate.HashBlock != this.lastCheckPoint.Hash)
```

Comment:
Perhaps this should be if the current block's height being validated matches the last checkpoint height, THEN check the hash?

@noescape00 can you confirm?

Reply:
This is already implicitly being checked by the preceding `if`s.

Reply:
ChainedHeaderToValidate here might be any block below last checkpoint. So most of the time it wont be equal to last checkpoint hash. 

Also no need to throw invalid header signature since it might be valid, just the block is not checkpointed. 

So probably just ignore the sig validation rule below last checkpoint, everything else is not needed.

Reply:
> ChainedHeaderToValidate here might be any block below last checkpoint. So most of the time it wont be equal to last checkpoint hash.

The code takes that into account by only checking the checkpoint block. Note that blocks following the last checkpoint block can only be deemed valid if we fully validate the checkpoint block when we encounter it. Not only its height must match but also its hash.

Reply:
Can you point me at which rule that is? I was trying to find a file named `XXXRule.cs` that contains "Checkpoint" and only found `PosCoinviewRule` - but that's not it, right?

Reply:
@quantumagi this is the code @noescape00 is referring to in CHT:

```
        private ConnectNewHeadersResult HandleCheckpointsHeader(ChainedHeader chainedHeader, ChainedHeader latestNewHeader, CheckpointInfo checkpoint, int peerId)
        {
            if (checkpoint.Hash != chainedHeader.HashBlock)
            {
                // Make sure that chain with invalid checkpoint in it is removed from the tree.
                // Otherwise a new peer may connect and present headers on top of invalid chain and we wouldn't recognize it.
                this.RemovePeerClaim(peerId, latestNewHeader);

                this.logger.LogDebug("Chained header '{0}' does not match checkpoint '{1}'.", chainedHeader, checkpoint.Hash);
                this.logger.LogTrace("(-)[INVALID_HEADER_NOT_MATCHING_CHECKPOINT]");
                throw new CheckpointMismatchException();
            }
```